### ISSUE TITLE
Type error when using SoftareaFunction forward

### ISSUE THREAD

**Issue by insuhan** [PerretB/ultrametric-fitting#2]
I got a type error when I call SoftareaFunction forward function. 

```
File "/WORKSPACE/ultrametric-fitting/ultrametric/softarea/softarea.py", line 57, in forward
    top_nodes)
TypeError: forward(): incompatible function arguments. The following argument types are supported:
    1. (arg0: hg::tree_internal::tree, arg1: numpy.ndarray[float64], arg2: hg::undirected_graph_internal::undirected_graph<hg::undirected_graph_internal::vecS>, arg3: numpy.nda
rray[int64], arg4: float, arg5: int) -> tuple
```

Looking inside softarea.cpp, the types for the first and second argument (const tree & tree, const ugraph& mst) raise the type error. It is suspicious that pybind11 conflicts the types defined in higra (e.g., tree, graph). Is this solvable by changing complie option? I am using higra 0.4.5 and python 3.7.5. 

Thanks in advance.

**Comment by PerretB** [comment]
Can you try again after upgrading Higra to the latest version (0.5.1) ? We recently solved some ABI compatibility issues that could make types compiled on your computer incompatible with types in the precompiled library (they regularly correct bugs in the ABI used by g++ that might break things).

**Comment by insuhan** [comment]
Thanks for the reply. I also test using Higra version 0.5.1 and 0.50 but the following error occurs just after call "from ultrametric.optimization import UltrametricFitting".

```
free(): invalid pointer
Aborted (core dumped)
```

Did you observe this error? or is the problem with my machine?

**Comment by PerretB** [comment]
I've never had such issue : if this appends during module loading, the c++ function is not even run at this moment so that might happen  during compiling or during dynamic module loading which I believe will run some code generated by pybind11. Can you give more precision on your system, compiler version, numpy/pytorch version ?

**Comment by insuhan** [comment]
Thanks for the quick reply. This is my machine setting.

Ubuntu 18.04.2 LTS
Python 3.7.5
numpy 1.17.4
pytorch 1.3.1
cuda 10.1
higra 0.5.1
pybind11 2.5.0
gcc 7.3.0

Thanks in advance!

**Comment by insuhan** [comment]
I found the reason. Higra should be imported after PyTorch. 

```
Python 3.7.5 (default, Oct 25 2019, 15:51:11)
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import higra, torch
free(): invalid pointer
Aborted (core dumped)
```

But when I execute "import torch, higra", it works fine. I think there might be an import issue of Higra conflicts with PyTorch. My current PyTorch version is 1.4.0 and Higra version is 0.5.1. Using the latest version of Higra, the compatibility issue is solved. Many thanks.

**Comment by PerretB** [comment]
Thank you for the solution. I don't understand what's happening but well :)

**Comment by PerretB** [comment]
Fixed in new Higra version 0.5.2 https://github.com/higra/Higra/commit/934babda91fc50a9e55bc5a740af3d9562fb7136
